### PR TITLE
Toggle element selector modal sidebar for small screens and zoom

### PR DESCRIPTION
**Pull request by gcamacho079**
### Description
- Updates the UI of element selector modals that have sidebars to include a sidebar toggle button when modal drops below a certain width (Resolves DEV-547)
- Updates `Craft.ui.createButton` function to use `html` and `id` from `config` object

Mobile view of a selector modal with a sidebar hidden behind a hamburger menu toggle
Screenshot of sidebar toggled open. A close button is at the top right of the sidebar, and the content is pushed offscreen while the sidebar is open

### CODE REVIEW

**Comment by gcamacho079** [comment]
Not quite ready for review yet, fixing a couple of things

File: src/web/assets/cp/src/js/BaseElementSelectorModal.js [hunk 1]
```diff
@@ -76,6 +84,136 @@ Craft.BaseElementSelectorModal = Garnish.Modal.extend(
...
+      if (footerHeight !== bottomPadding) {
+        this.$container.css('padding-bottom', footerHeight);
+      }
+    },
+
+    updateSidebarView: function () {
+      if (!this.supportSidebarToggleView) return;
+
+      if (this.sidebarShouldBeHidden()) {
+        if (!this.$sidebarToggleBtn) this.buildSidebarToggleView();
```
Comment (brandonkelly) [inline src/web/assets/cp/src/js/BaseElementSelectorModal.js:100]: Should probably be:
```suggestion
        if (!this.$sidebarToggleBtn || !this.$sidebarToggleBtn.length) {
            this.buildSidebarToggleView();
        }
```

File: src/web/assets/cp/src/js/BaseElementSelectorModal.js [hunk 2]
```diff
@@ -76,6 +84,136 @@ Craft.BaseElementSelectorModal = Garnish.Modal.extend(
...
+      } else {
+        if (this.$sidebarToggleBtn) this.resetView();
```
Comment (brandonkelly) [inline src/web/assets/cp/src/js/BaseElementSelectorModal.js:102]: Should probably be:

```suggestion
        if (this.$sidebarToggleBtn && this.$sidebarToggleBtn.length) {
            this.resetView();
        }
```

File: src/web/assets/cp/src/js/BaseElementSelectorModal.js [hunk 3]
```diff
@@ -76,6 +84,136 @@ Craft.BaseElementSelectorModal = Garnish.Modal.extend(
...
+      }
+    },
+
+    sidebarShouldBeHidden: function () {
+      const contentWidth = this.$container.outerWidth();
+      return contentWidth < 550;
+    },
+
+    resetView: function () {
+      if (this.$mainHeader) {
```
Comment (brandonkelly) [inline src/web/assets/cp/src/js/BaseElementSelectorModal.js:112]: Should probably be:

```suggestion
      if (this.$mainHeader && this.$mainHeader.length) {
```

File: src/web/assets/cp/src/js/BaseElementSelectorModal.js [hunk 4]
```diff
@@ -76,6 +84,136 @@ Craft.BaseElementSelectorModal = Garnish.Modal.extend(
...
+        this.$mainHeader.remove();
+      }
+
+      if (this.$sidebarHeader) {
```
Comment (brandonkelly) [inline src/web/assets/cp/src/js/BaseElementSelectorModal.js:116]: Should probably be:

```suggestion
      if (this.$sidebarHeader && this.$sidebarHeader.length) {
```

File: src/web/assets/cp/src/js/BaseElementSelectorModal.js [hunk 5]
```diff
@@ -76,6 +84,136 @@ Craft.BaseElementSelectorModal = Garnish.Modal.extend(
...
+
+        // If the focus is currently inside the sidebar, refocus the toggle
+        const $focusedEl = Garnish.getFocusedElement();
+        if ($.contains(this.$sidebar.get(0), $focusedEl.get(0)))
+          this.$sidebarToggleBtn.focus();
+      });
+    },
+
+    closeSidebar: function () {
+      if (!this.$sidebarToggleBtn) return;
```
Comment (brandonkelly) [inline src/web/assets/cp/src/js/BaseElementSelectorModal.js:201]: Should probably be:

```suggestion
      if (!this.$sidebarToggleBtn || !this.$sidebarToggleBtn.length) {
          return;
      }
```

File: src/web/assets/cp/src/js/BaseElementSelectorModal.js [hunk 6]
```diff
@@ -93,6 +231,17 @@ Craft.BaseElementSelectorModal = Garnish.Modal.extend(
       this.updateSelectBtnState();
     },
 
+    onSelectSource: function () {
+      this.updateHeading();
+      this.updateModalBottomPadding();
+    },
+
+    updateHeading: function () {
+      if (!this.$mainHeading) return;
```
Comment (brandonkelly) [inline src/web/assets/cp/src/js/BaseElementSelectorModal.js:240]: Should probably be:

```suggestion
      if (!this.$mainHeading || !this.$mainHeading.length) {
          return;
      }
```

**Review by brandonkelly — CHANGES_REQUESTED** [review]

File: src/web/assets/cp/src/js/BaseElementSelectorModal.js [hunk 7]
```diff
@@ -76,6 +84,136 @@ Craft.BaseElementSelectorModal = Garnish.Modal.extend(
...
+      }
+
+      this.$sidebarToggleBtn = null;
+      this.$body.addClass('has-sidebar');
+      this.$content.addClass('has-sidebar');
+      this.$sidebar.removeClass('hidden');
+    },
+
+    buildSidebarToggleView: function () {
+      if (this.$sidebarToggleBtn || !this.sidebarShouldBeHidden()) return;
```
[the remark this answers is not in the document]
Comment (gcamacho079) [reply src/web/assets/cp/src/js/BaseElementSelectorModal.js:127]: @brandonkelly question about this update. I'm not using the jQuery `find` method to set this variable, so `this.$sidebarToggleBtn` should only ever be `null` or the element (as created via the `Craft.ui.createButton` method. Should I still check the length of the jQuery object in instances like this?

[the remark this answers is not in the document]
Comment (brandonkelly) [reply src/web/assets/cp/src/js/BaseElementSelectorModal.js:127]: Sorry yep this one is fine as-is!

**Comment by gcamacho079** [comment]
> this one [`this.$sidebarToggleBtn`] is fine as-is!

`this.$mainHeading`, `this.$mainHeader`, `this.$sidebarHeader` are not set using the `find` method so I think the PR is good to go without the requested changes, but let me know if I misunderstood!

**Pull request merged**